src/bootstrap/src/core/build_steps/llvm.rs

```diff
@@ -242,15 +241,29 @@ pub(crate) fn is_ci_llvm_available(config: &Config, asserts: bool) -> bool {
 
 /// Returns true if we're running in CI with modified LLVM (and thus can't download it)
 pub(crate) fn is_ci_llvm_modified(config: &Config) -> bool {
-    CiEnv::is_rust_lang_managed_ci_job() && config.rust_info.is_managed_git_subrepository() && {
-        // We assume we have access to git, so it's okay to unconditionally pass
-        // `true` here.
-        let llvm_sha = detect_llvm_sha(config, true);
-        let head_sha =
-            output(helpers::git(Some(&config.src)).arg("rev-parse").arg("HEAD").as_command_mut());
-        let head_sha = head_sha.trim();
-        llvm_sha == head_sha
+    // If not running in a CI environment, return false.
+    if !CiEnv::is_ci() {
+        return false;
+    }
+
+    // In rust-lang/rust managed CI, assert the existence of the LLVM submodule.
+    if CiEnv::is_rust_lang_managed_ci_job() {
+        assert!(
+            config.in_tree_llvm_info.is_managed_git_subrepository(),
+            "LLVM submodule must be fetched in rust-lang/rust managed CI builders."
+        );
     }
+    // If LLVM submodule isn't present, skip the change check as it won't work.
+    else if !config.in_tree_llvm_info.is_managed_git_subrepository() {
+        return false;
+    }
+
+    let llvm_sha = detect_llvm_sha(config, true);
```

Comment:
I think that this is wrong, and [this](https://github.com/rust-lang/rust/pull/131305#discussion_r1794717517) code (which was in a previous commit) should be used instead.

Or at the very least, we should change the documentation of `get_closest_merge_commit`. Because `detect_llvm_sha` uses `get_closest_merge_commit`, which is supposed to return the latest **upstream** merge commit (that modified LLVM). In that case, it can by definition never return `HEAD` on CI (because `HEAD` is definitely not *upstream*).

However, we know that currently `get_closest_merge_commit` is bugged and probably just returns `HEAD` on CI, so we should probably first resolve https://github.com/rust-lang/rust/pull/131358 before continuing with this PR.

Reply:
This PR unblocks LLVM update PRs and it doesn't do anything other than improving current logic. I don't think #131358 should block this.

Reply:
Ok, let's fix the `download_ci_llvm` test for now to unblock the update PRs. But the logic is still wrong, https://github.com/rust-lang/rust/pull/131358 should hopefully fix it later.